### PR TITLE
Use generics in RunGetManyQueryForSchema

### PR DESCRIPTION
**Pull request by janisz**
## Description

After switching to Go 1.18 we can use generics. This will allow us to not parse output from database to `[][]byte` and then unmarshal it to `[]storage.*` but immediately parse to `[]storage.*`. So no intermediate slice will be created and we can save up some memory and allocations. 

Below tests shows comparison with previous commit not master.

```
name                   old time/op    new time/op    delta
GetMany/1_alerts-8        161µs ±11%     157µs ± 4%     ~     (p=0.065 n=19+19)
GetMany/2_alerts-8        196µs ± 8%     196µs ± 4%     ~     (p=0.708 n=19+20)
GetMany/4_alerts-8        238µs ± 7%     237µs ± 5%     ~     (p=0.687 n=19+20)
GetMany/8_alerts-8        338µs ± 7%     310µs ± 8%   -8.16%  (p=0.000 n=19+17)
GetMany/16_alerts-8       559µs ± 4%     470µs ± 6%  -15.79%  (p=0.000 n=18+20)
GetMany/32_alerts-8      1.00ms ±14%    0.79ms ± 4%  -21.21%  (p=0.000 n=19+20)
GetMany/64_alerts-8      2.10ms ±14%    1.71ms ± 5%  -18.51%  (p=0.000 n=17+19)
GetMany/128_alerts-8     4.57ms ±62%    2.93ms ± 7%  -35.88%  (p=0.000 n=20+20)
GetMany/256_alerts-8     7.66ms ± 8%    5.28ms ± 5%  -30.99%  (p=0.000 n=19+20)
GetMany/512_alerts-8     14.2ms ± 9%    10.1ms ± 4%  -28.77%  (p=0.000 n=20+20)
GetMany/1024_alerts-8    27.2ms ±15%    19.1ms ± 4%  -29.99%  (p=0.000 n=19+18)
GetMany/2048_alerts-8    51.9ms ± 5%    37.4ms ± 4%  -27.95%  (p=0.000 n=18+19)

name                   old alloc/op   new alloc/op   delta
GetMany/1_alerts-8       10.4kB ± 0%    10.4kB ± 0%   -0.16%  (p=0.000 n=20+20)
GetMany/2_alerts-8       18.3kB ± 0%    18.3kB ± 0%   -0.39%  (p=0.000 n=20+20)
GetMany/4_alerts-8       34.1kB ± 0%    33.9kB ± 0%   -0.54%  (p=0.000 n=18+20)
GetMany/8_alerts-8       66.4kB ± 0%    66.0kB ± 0%   -0.61%  (p=0.000 n=20+20)
GetMany/16_alerts-8       132kB ± 0%     130kB ± 0%   -1.01%  (p=0.000 n=20+20)
GetMany/32_alerts-8       263kB ± 0%     260kB ± 0%   -1.24%  (p=0.000 n=19+20)
GetMany/64_alerts-8       526kB ± 0%     519kB ± 0%   -1.37%  (p=0.000 n=20+20)
GetMany/128_alerts-8     1.05MB ± 0%    1.04MB ± 0%   -1.43%  (p=0.000 n=20+20)
GetMany/256_alerts-8     2.11MB ± 0%    2.08MB ± 0%   -1.44%  (p=0.000 n=20+19)
GetMany/512_alerts-8     4.21MB ± 0%    4.15MB ± 0%   -1.45%  (p=0.000 n=20+20)
GetMany/1024_alerts-8    8.49MB ± 0%    8.34MB ± 0%   -1.67%  (p=0.000 n=20+19)
GetMany/2048_alerts-8    16.9MB ± 0%    16.6MB ± 0%   -1.45%  (p=0.000 n=20+20)

name                   old allocs/op  new allocs/op  delta
GetMany/1_alerts-8          149 ± 0%       149 ± 0%     ~     (all equal)
GetMany/2_alerts-8          245 ± 0%       244 ± 0%   -0.41%  (p=0.000 n=20+20)
GetMany/4_alerts-8          437 ± 0%       434 ± 0%   -0.69%  (p=0.000 n=20+20)
GetMany/8_alerts-8          821 ± 0%       814 ± 0%   -0.85%  (p=0.000 n=20+20)
GetMany/16_alerts-8       1.59k ± 0%     1.57k ± 0%   -1.01%  (p=0.000 n=20+20)
GetMany/32_alerts-8       3.12k ± 0%     3.08k ± 0%   -1.08%  (p=0.000 n=20+18)
GetMany/64_alerts-8       6.18k ± 0%     6.11k ± 0%   -1.12%  (p=0.000 n=20+20)
GetMany/128_alerts-8      12.3k ± 0%     12.1k ± 0%   -1.09%  (p=0.000 n=20+20)
GetMany/256_alerts-8      24.5k ± 0%     24.2k ± 0%   -1.08%  (p=0.000 n=20+19)
GetMany/512_alerts-8      48.9k ± 0%     48.4k ± 0%   -1.08%  (p=0.000 n=20+19)
GetMany/1024_alerts-8     97.7k ± 0%     96.7k ± 0%   -1.08%  (p=0.000 n=19+20)
GetMany/2048_alerts-8      195k ± 0%      193k ± 0%   -1.08%  (p=0.000 n=20+15)

```

## Testing Performed

CI

### CODE REVIEW

**Comment by janisz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3446** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3446" target="_blank">Graphite</a> 
    * **PR #3448** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3448" target="_blank">Graphite</a> 
      * **PR #3449** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3449" target="_blank">Graphite</a> 
        * **PR #3450** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3450" target="_blank">Graphite</a> 
          * **PR #3451** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3451" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/3451?utm_source=stack-comment).

**Review by misberner — APPROVED** [review]

**Comment by misberner** [comment]
Nice, the solution with the `interface { *T ; ... }` is really non-obvious.

File: pkg/search/postgres/common.go [hunk 1]
```diff
@@ -753,6 +754,53 @@ func retryableRunGetManyQueryForSchema(ctx context.Context, query *query, db *pg
 	return results, nil
 }
 
+// RunGetManyQueryForSchemaType executes a request for just the search against the database and unmarshal it to given type.
```
Comment (misberner) [inline pkg/search/postgres/common.go:757]: Actually, coming back here, there's also a bunch of code duplication here. Maybe rewrite `retryableRunGetManyQueryForSchema` in a callback fashion, taking a `callback func([]byte) error`? This should avoid the `[][]byte` allocation in the proto case. Note that you _need_ to shift it to the inner `retryable...` function, as callbacks and retries don't go too well together.

**Review by misberner — CHANGES_REQUESTED** [review]

File: pkg/search/postgres/common.go (hunk 1 above)
Comment (janisz) [reply pkg/search/postgres/common.go:757]: I remove this functions here https://github.com/stackrox/stackrox/pull/3454
Let me merge this PRs so diff will be easier to read

File: pkg/search/postgres/common.go [hunk 2]
```diff
@@ -846,3 +845,27 @@ func redactedQueryData(query *query) interface{} {
...
+	var data []byte
+	if err := row.Scan(&data); err != nil {
+		return nil, errors.Wrap(err, "scanning row")
+	}
+	msg := PT(new(T))
+	if err := msg.Unmarshal(data); err != nil {
+		return nil, errors.Wrapf(err, "unmarshaling %T", msg)
+	}
+	return msg, nil
+}
```
Comment (misberner) [inline pkg/search/postgres/common.go:871]: Super inconsequential, but I think this should rather return a `*T`. You can even do with only a single explicit cast:
```suggestion
func unmarshal[T any, PT unmarshaler[T]](row pgx.Row) (*T, error) {
	var data []byte
	if err := row.Scan(&data); err != nil {
		return nil, errors.Wrap(err, "scanning row")
	}
	msg := new(T)
	if err := PT(msg).Unmarshal(data); err != nil {
		return nil, errors.Wrapf(err, "unmarshaling %T", msg)
	}
	return msg, nil
}
```

**Review by misberner — APPROVED** [review]

File: pkg/search/postgres/common.go [hunk 3]
```diff
@@ -846,3 +845,27 @@ func redactedQueryData(query *query) interface{} {
...
 }
+
+func scanRows[T any, PT unmarshaler[T]](rows pgx.Rows) ([]*T, error) {
+	var results []*T
+	for rows.Next() {
+		msg, err := unmarshal[T, PT](rows)
+		if err != nil {
+			return nil, err
+		}
+		results = append(results, (*T)(msg))
```
Comment (misberner) [inline pkg/search/postgres/common.go:856]:
```suggestion
		results = append(results, msg)
```

**Review by misberner — APPROVED** [review]

**Review by md2119 — APPROVED** [review]
Noted that you updated the hand-rolled image store. Node store changes seems to be missing.

**Pull request merged**